### PR TITLE
Improve Error Handling in TreeTime

### PR DESCRIPTION
**Pull request by anna-parker**
# Related Issue and PR:

https://github.com/nextstrain/augur/issues/1032
https://github.com/nextstrain/augur/pull/1033

Building on @victorlin's suggestions in : https://github.com/neherlab/treetime/compare/44e7c56a1d10baa3d4dcdf07b5c7cd55d6d651e9...2632d96144ada719a33f966cc11c5bc118c987a4 I have made a couple minor changes to TreeTime's error handling process: 
 - more consistent labeling of error classes: rename errors due to incorrect input data or incorrect use of treetime as one of the three predefined TreeTime error classes: `MissingDataError`, `UnknownMethodError` and `NotReadyError`
 - add new `TreeTimeError` class for errors that might be due to input data not matching with assumptions or unknown behavior (potential bug): `TreeTimeOtherError`. 
 - call `treetime.run` in a wrapper to catch all errors with notification on how to report issues. i.e. `ERROR in TreeTime.run: An error has occurred which is not properly handled in TreeTime. If this error persists, please let us know "
                    "by filing a new issue including the original command and the error above at: https://github.com/neherlab/treetime/issues `
 - when run from another application, such as `augur` return all errors as `TreeTimeError`s but with distinction between input/usage errors and potential treetime bugs (these are output as `TreeTimeOtherError`s). On standard run print traceback for unknown errors / potential bugs, print warning and then exit treetime (avoid double printing of traceback or traceback after final warning). 
 
 Example for unknown `ValueError`: 
 
image

Example for known `MissingDataError`: 

image

I used the same example as in the PR above to create an error. I made two test cases, one where the error is called a `ValueError` which should be handled as a `TreeTimeOtherError` and another where the error is called a `MissingDataError`. 

# Testing
- install augur and treetime in local environment
- cause an Exception in treetime (either call this a `ValueError` or a `TreeTimeError`), I added a line in `_exp_lt`
```
        """
        Parameters
        ----------

         t : float
            time to propagate

        Returns
        --------

         exp_lt : numpy.array
            Array of values exp(lambda(i) * t),
            where (i) - alphabet index (the eigenvalue number).
        """
        log_val = self.mu * t * self.eigenvals
        log_val = np.ones(len(log_val))*750 ##cause an exception
        if any(i > 10 for i in log_val):
            raise ValueError("Error in computing exp(Q * t): Q has positive eigenvalues or the branch length t \n"
                    "is too large. This is most likely caused by incorrect input data.")

        return np.exp(log_val)
```
- install modified treetime in local environment `(pip install .`)
- call `treetime.run()` in a test script where it is not in a `try` - `catch` statement, for example: run `python test/treetime_algorithms.py`

### CODE REVIEW

File: treetime/__init__.py [hunk 1]
```diff
@@ -1,5 +1,8 @@
 version="0.9.2"
-
+## Here we define an error class for TreeTime errors, MissingData, UnknownMethod and NotReady errors
+## are all due to incorrect calling of TreeTime functions or input data that does not fit our base assumptions.
+## Errors marked as TreeTimeOtherErrors might be due to data not fulfilling base assumptions or due 
+## to bugs in TreeTime. Please report them to the developers if they persist. 
 class TreeTimeError(Exception):
     """TreeTimeError class"""
```
Comment (corneliusroemer) [inline treetime/__init__.py:7]:
```suggestion
    """
    TreeTimeError class
    Parent class for more specific errors
    Raised when treetime is used incorrectly in contrast with `TreeTimeOtherError`
    `TreeTimeOtherError` is raised when the reason of the error is unknown, could indicate bug     
    """
```

File: treetime/__init__.py [hunk 2]
```diff
@@ -9,13 +12,17 @@ class MissingDataError(TreeTimeError):
     pass
```
Comment (corneliusroemer) [inline treetime/__init__.py:12]: Or parameters are missing? If I understand usage below correctly?

File: treetime/clock_tree.py [hunk 3]
```diff
@@ -77,7 +77,7 @@ def __init__(self, *args, dates=None, debug=False, real_dates=True, precision_ff
         """
         super(ClockTree, self).__init__(*args, **kwargs)
         if dates is None:
-            raise ValueError("ClockTree requires date constraints!")
+            raise MissingDataError("ClockTree requires date constraints!")
```
Comment (corneliusroemer) [inline treetime/clock_tree.py:80]: If this is how you want to use MissingDataError one should probably change the description of the MissingDataError to encompass missing parameters

File: treetime/clock_tree.py [hunk 4]
```diff
@@ -203,7 +203,7 @@ def _set_precision_fft(self, precision_fft, precision_branch='auto'):
                         " fft_grid_points=%.3e"%(precision_fft), 2)
                 self.fft_grid_size = precision_fft
             elif precision_fft!='auto':
-                raise ValueError(f"ClockTree: precision_fft needs to be either 'auto' or an integer, got '{precision_fft}'.")
+                raise UnknownMethodError(f"ClockTree: precision_fft needs to be either 'auto' or an integer, got '{precision_fft}'.")
```
Comment (corneliusroemer) [inline treetime/clock_tree.py:206]: This should be a new type of error? Something like: Wrong parameter passed - not unknown method. This has nothing to do with a method?

**Review by corneliusroemer — CHANGES_REQUESTED** [review]
Still a bit confused what the error hierarchy is supposed to be :)

File: treetime/treetime.py [hunk 5]
```diff
@@ -51,7 +51,29 @@ def __init__(self, *args,**kwargs):
         super(TreeTime, self).__init__(*args, **kwargs)
 
 
-    def run(self, root=None, infer_gtr=True, relaxed_clock=None, n_iqd = None,
+    def run(self, augur=False, **kwargs):
```
Comment (anna-parker) [inline treetime/treetime.py:54]: @victorlin Hi, I modified how the `run` function is called to make sure that the Errors get printed at the correct location when running in `TreeTime` and in `augur` - however @corneliusroemer mentioned that using the argument `augur=true` is not the best coding practice - is there a better way to do this? Perhaps we could use @rneher `raise_uncaught_exception = true` instead as an argument to make it less specific and then add some commentary? Thanks for any input :-)

**Review by victorlin — COMMENTED** [review]
Post-merge requested changes

**Comment by anna-parker** [comment]
@victorlin thanks so much for your suggestions - I have added them to another branch https://github.com/neherlab/treetime/tree/feat/victor-error-handling - will do a PR after checking they integrate with your suggestions in augur :-)

File: treetime/__init__.py [hunk 6]
```diff
@@ -1,21 +1,33 @@
...
 class UnknownMethodError(TreeTimeError):
-    """MissingDataError class raised when tree or alignment are missing"""
+    """MissingDataError class raised when an unknown method is called"""
     pass
 
 class NotReadyError(TreeTimeError):
     """NotReadyError class raised when results are requested before inference"""
     pass
 
+class TreeTimeOtherError(Exception):
```
[the remark this answers is not in the document]
Comment (corneliusroemer) [reply treetime/__init__.py:27]: I like unknown

**Pull request merged**